### PR TITLE
fix: Schema format and validation fixes. 

### PR DESCRIPTION
**Pull request by Bento007**
## Reason for Change

- NCBITaxon, CL, EFO,  and HANCESTRO were failing schema check because they had terms that have less than 7 digits.
- Unified the json schema regex into a single location to make keep them up to date easier.
- all_ontoligies was not generating json that matches the expected schema.
- Fixed an issue with new json.gz files being committed when no changes to the data were made. This was happening because timestamp of the zipped file keeps changing, resulting in what appear to be a modified file. 
- There was an issue with schema validation failing on the first error and not validating the remaining schemas. This has been fixed to make fixing validation errors easier.

## Changes

- centralize the regexes used to identify ontology term ids to a central location
- fixed a typo in all_ontology_schema.json#$schema
- append pytest fixture with `_fixtures` when a none mock values is returned.
- all ontology terms can have `[0-9]+` digit in their term_id
- time stamp the gzip file to 0 so only changes to the data will result in a new gzip files being committed
- add validation after generating all_ontology files. Do not commit if they don't pass validation
- Modify schema validation logic to validate all files before failing with errors.

## Testing steps

- ran validate_scheme on existing json.
- ran generate_all_ontologies locally and in GHA

### CODE REVIEW

**Comment by Bento007** [comment]
@nayib-jose-gloria when I generate all_ontologies, I see schema mismatches n the "replace_by" terms. Should we be supporting these terms here? 
```
Processing /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/EFO-ontology-v3.62.0.json.gz
ERROR:validate_json_schemas:Error validating /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/EFO-ontology-v3.62.0.json.gz against /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/artifact-schemas/all_ontology_schema.json
Traceback (most recent call last):
  File "/Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/tools/ontology-builder/src/validate_json_schemas.py", line 71, in verify_json
    validate(instance=data, schema=schema, registry=registry)
  File "/Users/trentsmith/workspace/test-cellxgene-ontology/venv-api/lib/python3.11/site-packages/jsonschema/validators.py", line 1312, in validate
    raise error
jsonschema.exceptions.ValidationError: 'PO:0009088' is not valid under any of the given schemas

Failed validating 'anyOf' in schema['additionalProperties']['properties']['replaced_by']:
    {'anyOf': [{'$ref': '#/definitions/CL_term_id'},
               {'$ref': '#/definitions/EFO_term_id'},
               {'$ref': '#/definitions/HANCESTRO_term_id'},
               {'$ref': '#/definitions/HsapDv_term_id'},
               {'$ref': '#/definitions/MmusDv_term_id'},
               {'$ref': '#/definitions/MONDO_term_id'},
               {'$ref': '#/definitions/NCBITaxon_term_id'},
               {'$ref': '#/definitions/PATO_term_id'},
               {'$ref': '#/definitions/UBERON_term_id'}]}

On instance['EFO:0001021']['replaced_by']:
    'PO:0009088'
Processing /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/HANCESTRO-ontology-3.0.json.gz
Processing /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/CL-ontology-v2024-01-04.json.gz
* Owlready2 * Warning: ignoring cyclic subclass of/subproperty of, involving:
  http://purl.obolibrary.org/obo/CHEBI_36080
  http://purl.obolibrary.org/obo/PR_000000001

ERROR:validate_json_schemas:Error validating /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/CL-ontology-v2024-01-04.json.gz against /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/artifact-schemas/all_ontology_schema.json
Traceback (most recent call last):
  File "/Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/tools/ontology-builder/src/validate_json_schemas.py", line 71, in verify_json
    validate(instance=data, schema=schema, registry=registry)
  File "/Users/trentsmith/workspace/test-cellxgene-ontology/venv-api/lib/python3.11/site-packages/jsonschema/validators.py", line 1312, in validate
    raise error
jsonschema.exceptions.ValidationError: 'PO:0000332' is not valid under any of the given schemas

Failed validating 'anyOf' in schema['additionalProperties']['properties']['replaced_by']:
    {'anyOf': [{'$ref': '#/definitions/CL_term_id'},
               {'$ref': '#/definitions/EFO_term_id'},
               {'$ref': '#/definitions/HANCESTRO_term_id'},
               {'$ref': '#/definitions/HsapDv_term_id'},
               {'$ref': '#/definitions/MmusDv_term_id'},
               {'$ref': '#/definitions/MONDO_term_id'},
               {'$ref': '#/definitions/NCBITaxon_term_id'},
               {'$ref': '#/definitions/PATO_term_id'},
               {'$ref': '#/definitions/UBERON_term_id'}]}

On instance['CL:0000400']['replaced_by']:
    'PO:0000332'
Processing /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/HsapDv-ontology-11.json.gz
Processing /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/PATO-ontology-v2023-05-18.json.gz
* Owlready2 * Warning: ignoring cyclic subclass of/subproperty of, involving:
  http://purl.obolibrary.org/obo/BFO_0000023
  http://purl.obolibrary.org/obo/CHEBI_50906

* Owlready2 * Warning: ignoring cyclic subclass of/subproperty of, involving:
  http://purl.obolibrary.org/obo/CHEBI_36080
  http://purl.obolibrary.org/obo/PR_000000001

Processing /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/NCBITaxon-ontology-v2023-06-20.json.gz
```

**Comment by nayib-jose-gloria** [comment]
> @nayib-jose-gloria when I generate all_ontologies, I see schema mismatches n the "replace_by" terms. Should we be supporting these terms here?
> 
> ```
> Processing /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/EFO-ontology-v3.62.0.json.gz
> ERROR:validate_json_schemas:Error validating /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/EFO-ontology-v3.62.0.json.gz against /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/artifact-schemas/all_ontology_schema.json
> Traceback (most recent call last):
>   File "/Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/tools/ontology-builder/src/validate_json_schemas.py", line 71, in verify_json
>     validate(instance=data, schema=schema, registry=registry)
>   File "/Users/trentsmith/workspace/test-cellxgene-ontology/venv-api/lib/python3.11/site-packages/jsonschema/validators.py", line 1312, in validate
>     raise error
> jsonschema.exceptions.ValidationError: 'PO:0009088' is not valid under any of the given schemas
> 
> Failed validating 'anyOf' in schema['additionalProperties']['properties']['replaced_by']:
>     {'anyOf': [{'$ref': '#/definitions/CL_term_id'},
>                {'$ref': '#/definitions/EFO_term_id'},
>                {'$ref': '#/definitions/HANCESTRO_term_id'},
>                {'$ref': '#/definitions/HsapDv_term_id'},
>                {'$ref': '#/definitions/MmusDv_term_id'},
>                {'$ref': '#/definitions/MONDO_term_id'},
>                {'$ref': '#/definitions/NCBITaxon_term_id'},
>                {'$ref': '#/definitions/PATO_term_id'},
>                {'$ref': '#/definitions/UBERON_term_id'}]}
> 
> On instance['EFO:0001021']['replaced_by']:
>     'PO:0009088'
> Processing /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/HANCESTRO-ontology-3.0.json.gz
> Processing /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/CL-ontology-v2024-01-04.json.gz
> * Owlready2 * Warning: ignoring cyclic subclass of/subproperty of, involving:
>   http://purl.obolibrary.org/obo/CHEBI_36080
>   http://purl.obolibrary.org/obo/PR_000000001
> 
> ERROR:validate_json_schemas:Error validating /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/CL-ontology-v2024-01-04.json.gz against /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/artifact-schemas/all_ontology_schema.json
> Traceback (most recent call last):
>   File "/Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/tools/ontology-builder/src/validate_json_schemas.py", line 71, in verify_json
>     validate(instance=data, schema=schema, registry=registry)
>   File "/Users/trentsmith/workspace/test-cellxgene-ontology/venv-api/lib/python3.11/site-packages/jsonschema/validators.py", line 1312, in validate
>     raise error
> jsonschema.exceptions.ValidationError: 'PO:0000332' is not valid under any of the given schemas
> 
> Failed validating 'anyOf' in schema['additionalProperties']['properties']['replaced_by']:
>     {'anyOf': [{'$ref': '#/definitions/CL_term_id'},
>                {'$ref': '#/definitions/EFO_term_id'},
>                {'$ref': '#/definitions/HANCESTRO_term_id'},
>                {'$ref': '#/definitions/HsapDv_term_id'},
>                {'$ref': '#/definitions/MmusDv_term_id'},
>                {'$ref': '#/definitions/MONDO_term_id'},
>                {'$ref': '#/definitions/NCBITaxon_term_id'},
>                {'$ref': '#/definitions/PATO_term_id'},
>                {'$ref': '#/definitions/UBERON_term_id'}]}
> 
> On instance['CL:0000400']['replaced_by']:
>     'PO:0000332'
> Processing /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/HsapDv-ontology-11.json.gz
> Processing /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/PATO-ontology-v2023-05-18.json.gz
> * Owlready2 * Warning: ignoring cyclic subclass of/subproperty of, involving:
>   http://purl.obolibrary.org/obo/BFO_0000023
>   http://purl.obolibrary.org/obo/CHEBI_50906
> 
> * Owlready2 * Warning: ignoring cyclic subclass of/subproperty of, involving:
>   http://purl.obolibrary.org/obo/CHEBI_36080
>   http://purl.obolibrary.org/obo/PR_000000001
> 
> Processing /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/NCBITaxon-ontology-v2023-06-20.json.gz
> Processing /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/MmusDv-ontology-9.json.gz
> Processing /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/MONDO-ontology-v2024-01-03.json.gz
> ERROR:validate_json_schemas:Error validating /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/MONDO-ontology-v2024-01-03.json.gz against /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/artifact-schemas/all_ontology_schema.json
> Traceback (most recent call last):
>   File "/Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/tools/ontology-builder/src/validate_json_schemas.py", line 71, in verify_json
>     validate(instance=data, schema=schema, registry=registry)
>   File "/Users/trentsmith/workspace/test-cellxgene-ontology/venv-api/lib/python3.11/site-packages/jsonschema/validators.py", line 1312, in validate
>     raise error
> jsonschema.exceptions.ValidationError: 'HP:0001115' is not valid under any of the given schemas
> 
> Failed validating 'anyOf' in schema['additionalProperties']['properties']['replaced_by']:
>     {'anyOf': [{'$ref': '#/definitions/CL_term_id'},
>                {'$ref': '#/definitions/EFO_term_id'},
>                {'$ref': '#/definitions/HANCESTRO_term_id'},
>                {'$ref': '#/definitions/HsapDv_term_id'},
>                {'$ref': '#/definitions/MmusDv_term_id'},
>                {'$ref': '#/definitions/MONDO_term_id'},
>                {'$ref': '#/definitions/NCBITaxon_term_id'},
>                {'$ref': '#/definitions/PATO_term_id'},
>                {'$ref': '#/definitions/UBERON_term_id'}]}
> 
> On instance['MONDO:0000348']['replaced_by']:
>     'HP:0001115'
> Processing /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/UBERON-ontology-v2024-01-18.json.gz
> ERROR:validate_json_schemas:Error validating /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/ontology-assets/UBERON-ontology-v2024-01-18.json.gz against /Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/artifact-schemas/all_ontology_schema.json
> Traceback (most recent call last):
>   File "/Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/tools/ontology-builder/src/validate_json_schemas.py", line 71, in verify_json
>     validate(instance=data, schema=schema, registry=registry)
>   File "/Users/trentsmith/workspace/test-cellxgene-ontology/venv-api/lib/python3.11/site-packages/jsonschema/validators.py", line 1312, in validate
>     raise error
> jsonschema.exceptions.ValidationError: 'ZFA:0001473' is not valid under any of the given schemas
> 
> Failed validating 'anyOf' in schema['additionalProperties']['properties']['replaced_by']:
>     {'anyOf': [{'$ref': '#/definitions/CL_term_id'},
>                {'$ref': '#/definitions/EFO_term_id'},
>                {'$ref': '#/definitions/HANCESTRO_term_id'},
>                {'$ref': '#/definitions/HsapDv_term_id'},
>                {'$ref': '#/definitions/MmusDv_term_id'},
>                {'$ref': '#/definitions/MONDO_term_id'},
>                {'$ref': '#/definitions/NCBITaxon_term_id'},
>                {'$ref': '#/definitions/PATO_term_id'},
>                {'$ref': '#/definitions/UBERON_term_id'}]}
> 
> On instance['UBERON:2001473']['replaced_by']:
>     'ZFA:0001473'
> Traceback (most recent call last):
>   File "/Users/trentsmith/workspace/test-cellxgene-ontology/cellxgene-ontology/tools/ontology-builder/src/all_ontology_generator.py", line 272, in <module>
>     raise Exception("Some ontology files do not match the schema")
> Exception: Some ontology files do not match the schema
> ```

I think we should support it for replaced_by terms; in the end, if the ontologies report it, we should provide that info to users and they can then choose to use or ignore that information accordingly (similar to how CL:0000003 is 'replaced by' CL:0000000 in the CL ontology, but we chose to use 'unknown' instead in our schema)

File: tools/ontology-builder/requirements.txt [hunk 1]
```diff
@@ -2,3 +2,5 @@ owlready2==0.45
 PyYaml==6.0.1
 jsonschema==4.21.1
 semantic-version==2.10.0
+referencing # used by jsonschema
```
Comment (nayib-jose-gloria) [inline tools/ontology-builder/requirements.txt:5]: can we pin?

**Comment by Bento007** [comment]
> I think we should support it for replaced_by terms; in the end, if the ontologies report it, we should provide that info to users and they can then choose to use or ignore that information accordingly (similar to how CL:0000003 is 'replaced by' CL:0000000 in the CL ontology, but we chose to use 'unknown' instead in our schema). It should ultimately be the validator that has opinions on what to use or ignore from the ontologies in a dataset. We only use this function in the dry-run report as of now. We can update the docstring to reflect that the ontologies could pick a term from a different ontology.

which docstring should we update?

**Review by nayib-jose-gloria — APPROVED** [review]

**Pull request merged**